### PR TITLE
Generate coverage reports using only GitHub Actions, without codecov.io

### LINKED ISSUE #410: Code coverage via GitHub action
This is an experiment.

I don't have time now to work on it, but I left a few comments/links

https://github.com/ewjoachim/python-coverage-comment-action

----------

The idea is to generate the coverage report only using GitHub Actions

This includes the commit status and comment with the report.

Basically, reinvent codecov.io

Some ideas here 

https://hynek.me/articles/ditch-codecov-python/

https://itnext.io/github-actions-code-coverage-without-third-parties-f1299747064d

https://josh-ops.com/posts/github-code-coverage/

https://dev.to/thejaredwilcurt/coverage-badge-with-github-actions-finally-59fa

### PR DESCRIPTION
**Pull request by adiroiban**
# Description

Fixes #410

This tries to generate coverage reports only via GitHub Actions...so that we don't have to depend on codecov.io

# Drive-by

I have removed the codeowner as I don't think it helps.

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (altendky) [inline .github/workflows/ci.yml]: Maybe also create an html report and artifact it for download?

File: .github/workflows/ci.yml
Comment (altendky) [inline .github/workflows/ci.yml:270]:
```suggestion
    - uses: actions/checkout@v3
      with:
        fetch-depth: 0
        
```

File: .github/workflows/ci.yml
Comment (adiroiban) [reply .github/workflows/ci.yml]: My plan is to send the diff coverage report as a comment... downloading the artifact looks complicated.

**Comment by adiroiban** [comment]
I think this is ready for an initial feedback

needs-review

File: .github/workflows/ci.yml
Comment (altendky) [reply .github/workflows/ci.yml]: Yeah, it's wouldn't be a primary use.  But, it is how you can you get a nice view of the code and coverage without having somewhere to upload the html.  Anyways, not critical, just a thought that can also be easily added later.

File: .github/workflows/ci.yml
Comment (altendky) [inline .github/workflows/ci.yml]: This is still needed after `fetch-depth: 0` above?

File: .github/workflows/ci.yml
Comment (altendky) [inline .github/workflows/ci.yml]: Not much point in updating pip if you don't do that first and then make a second pip call to install the other stuff.  Both options are available below as suggestions.

(I just realized I could add multiple comments in the same discussion at the same time so that this can work...)

File: .github/workflows/ci.yml
Comment (altendky) [inline .github/workflows/ci.yml]: Slight wording tweak, though I'm not sure there's really a need to have the run ID in the name.  Yes, the files need to be unique but AFAIK uploading like this uses names within a namespace for this run.  And actually, if you get a new run ID for reruns then I think this setup wouldn't work in case a job flakes and you rerun it as you would then get two separate artifacts.

```suggestion
        # We have a single artifact shared for each workflow run.
        # in this way we should not interfere with coverage reports from other PRs
        # or previous runs for the same PR.
```

File: .github/workflows/ci.yml
Comment (altendky) [inline .github/workflows/ci.yml]:
```suggestion
        # It is important to keep the unique names for the coverage files
        # so that when uploaded to the same artifact, they don't overlap.
```

File: .github/workflows/ci.yml
Comment (altendky) [inline .github/workflows/ci.yml]: See comments in duplicated section below.

Side note...  I really like cross-platform matrices to avoid this.  `:]`

File: noxfile.py
Comment (altendky) [inline noxfile.py]:
```suggestion
        # When running the test locally, show the coverage report.
```

File: pyproject.toml
Comment (altendky) [inline pyproject.toml]: I guess this config with slashes each direction is compensating for now for the Windows issues discussed in https://github.com/nedbat/coveragepy/issues/991.  I'd suggest just using the 6.6.0b1 that has a fix for that but...  https://github.com/pypa/pip/issues/11565  I'm not sure, towncrier may or may not trigger that bug.  So maybe best to just wait and leave a comment here as reference.

```suggestion
source = [
    "src",
    "*.nox/*/site-packages",
    # required until coverage 6.6.0 is used: https://github.com/nedbat/coveragepy/issues/991
    "*.nox\\*\\*\\site-packages"
]
```

File: .github/workflows/ci.yml
Comment (altendky) [reply .github/workflows/ci.yml]:
```suggestion
      run: python -m pip install --upgrade coverage[toml] diff_cover
```

Comment (altendky) [reply .github/workflows/ci.yml]:
```suggestion
      run: |
        python -m pip install --upgrade pip
        python -m pip install --upgrade coverage[toml] diff_cover
```

**Review by altendky — COMMENTED** [review]
I know this isn't part of the PR, but it's on topic so I'll at least comment.

https://github.com/twisted/towncrier/blob/a86b23b1e206cdf5897d59f6c72a1df8d227ce8c/pyproject.toml#L107-L110

Why would we omit tests?  Not-covered test code is also bad.

Anyways, I ended up with a lot of comments.  `:|`  I hope they are helpful.

File: .github/workflows/ci.yml
Comment (adiroiban) [reply .github/workflows/ci.yml]: I don't know. I think so. Let's see.

```suggestion
```

**Comment by adiroiban** [comment]
thanks @altendky  for the review. 

> Why would we omit tests? Not-covered test code is also bad.

I don't know. I have enabled the test coverage. Looks 100%

I also think that test code coverage is super important... you might think that you have tests covering the prod code but in fact the tests are always skipped.

**Comment by adiroiban** [comment]
> Maybe also create an html report and artifact it for download?

Searching for artifacts, downloading the artifacts, extracting the archive, and loading in a local browser... seems complicated.

For me is easier to just run the tests and check the local report.

---------

The idea with CI is just to raise a warning if coverage is missing.

File: .github/workflows/ci.yml
Comment (adiroiban) [reply .github/workflows/ci.yml]: I am missing the comment.

I am happy to see less duplication, but I think is best to have it in separate PR.

Thanks

File: .github/scripts/pr_comment.js
Comment (altendky) [inline .github/scripts/pr_comment.js]: Looks like a stray space got introduced in https://github.com/twisted/towncrier/pull/447/commits/8208990ecbdbd4454d39a3f461215c7aba748e62#diff-e24afedc0691665628968c647e1b47778e5678bcc154c1c7e83ab9b1d63680b2.

```suggestion
        var comment_body = body + comment_marker
```

**Comment by adiroiban** [comment]
Thanks Kyle for the review.

I think that this is ready for another review.

For now, the output/feedback is done via a single comment that is updated and contains the output from `coverage report` and `diff-cover`

File: .github/workflows/ci.yml
Comment (altendky) [reply .github/workflows/ci.yml]: This is copy pasted from the other similar section and contains the same typos that were present there and have been fixed.  They should be fixed here as well.

But, I'm going to go ahead and trigger a run here, cancel it partially completed, rerun the cancelled jobs and see if my concern about using the `github.run_id` seems accurate.

**Review by altendky — COMMENTED** [review]
Thanks for the cleanup, a couple things left to check on.

**Comment by adiroiban** [comment]
Re-run will keep the run-id. Is not ideal ... but I think it should work.

**Comment by altendky** [comment]
Well, I was trying over at https://github.com/twisted/towncrier/actions/runs/3401839419 and yes, the re-runs/attempts do retain the same run ID as you say.  So, what is the need for the run ID here?  Just so that if you download the artifact locally you know what run it corresponds to?

**Review by altendky — APPROVED** [review]
Thanks!

**Comment by adiroiban** [comment]
I added the run-id to make sure we don't mix coverages between commits or PRs

Based on the docs from here 

https://github.com/actions/upload-artifact#uploading-to-the-same-artifact

> Each artifact behaves as a file share. Uploading to the same artifact multiple times in the same workflow can overwrite and append already uploaded files:

But re-reading it I see "in the same workflow"

So I think that the run ID is not needed.

at first I thought that the artifacts are shared between multiple workflows and runs.

I will remove the run-id as it is not needed

**Comment by adiroiban** [comment]
Many thanks, Kyle for the review and useful feedback.

I will not merge it just now as I would like to apply the same process to twisted/twisted and apply any changes made there to keep the code in sync.

here is how it looks for twisted/twisted https://github.com/twisted/twisted/pull/11745

We still need a real PR with a real diff coverage...but we can see if after this is merged and update the format once we have a real example

**Comment by adiroiban** [comment]
I now see that diff-covers supports markdown output. I have enabled it.

diff-covers also supports json report so in theory we can parse that output and generate inline PR comments, similar to codecov.io

I will give it a try here.

----------

For me, the HTML artifact report doesn't look helpful.
I would like to have everything reported inside the PR via comments or inline diff comments/annotations.

So future work, if someting is missing I think is best to contribute to diff-cover so that the markdown or json reports will do what we want.


------------

For example, I see that `diff-cover` range detection can't merge ranges splited by a comment line.

```
coverage report
src/towncrier/build.py                  92      7     30      2    93%   123-126, 138-142
```

```
diff-cover
src/towncrier/build.py (22.2%): Missing lines 123,125-126,138,140-142
```

**Comment by adiroiban** [comment]
I will merge this an open a new PR in which I will try to get inline comments working.

**Comment by adiroiban** [comment]
A better example for the report is here https://github.com/twisted/towncrier/pull/451

**Comment by hynek** [comment]
should we maybe roll this back for now, since it's breaking on all PRs that aren't by us?

**Comment by adiroiban** [comment]
Yes. I think this should be reverted. @hynek  Will you do that?
Thanks!

**Pull request merged**
